server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/restore/RestoreClusterStateListener.java

```diff
@@ -11,84 +11,94 @@
+import org.elasticsearch.node.NodeClosedException;
 import org.elasticsearch.snapshots.RestoreInfo;
 import org.elasticsearch.snapshots.RestoreService;
 
 import java.util.Map;
-import java.util.function.Supplier;
 
 import static org.elasticsearch.snapshots.RestoreService.restoreInProgress;
 
-public class RestoreClusterStateListener implements ClusterStateListener {
+public class RestoreClusterStateListener {
 
     private static final Logger logger = LogManager.getLogger(RestoreClusterStateListener.class);
 
-    private final ClusterService clusterService;
-    private final String uuid;
-    private final ActionListener<RestoreSnapshotResponse> listener;
-    private final Supplier<ThreadContext.StoredContext> contextSupplier;
+    private RestoreClusterStateListener() {}
 
-    private RestoreClusterStateListener(
+    /**
+     * Creates a cluster state listener and registers it with the cluster service. The listener passed as a
+     * parameter will be called when the restore is complete.
+     */
+    public static void createAndRegisterListener(
         ClusterService clusterService,
         RestoreService.RestoreCompletionResponse response,
         ActionListener<RestoreSnapshotResponse> listener,
-        Supplier<ThreadContext.StoredContext> contextSupplier
+        ThreadContext threadContext
     ) {
-        this.clusterService = clusterService;
-        this.uuid = response.getUuid();
-        this.listener = listener;
-        this.contextSupplier = contextSupplier;
-    }
-
-    @Override
-    public void clusterChanged(ClusterChangedEvent changedEvent) {
-        try (ThreadContext.StoredContext stored = contextSupplier.get()) {
-            final RestoreInProgress.Entry prevEntry = restoreInProgress(changedEvent.previousState(), uuid);
-            final RestoreInProgress.Entry newEntry = restoreInProgress(changedEvent.state(), uuid);
-            if (prevEntry == null) {
-                // When there is a master failure after a restore has been started, this listener might not be registered
-                // on the current master and as such it might miss some intermediary cluster states due to batching.
-                // Clean up listener in that case and acknowledge completion of restore operation to client.
-                clusterService.removeListener(this);
-                listener.onResponse(new RestoreSnapshotResponse((RestoreInfo) null));
-            } else if (newEntry == null) {
-                clusterService.removeListener(this);
-                Map<ShardId, RestoreInProgress.ShardRestoreStatus> shards = prevEntry.shards();
-                assert prevEntry.state().completed() : "expected completed snapshot state but was " + prevEntry.state();
+        final String uuid = response.getUuid();
+        final DiscoveryNode localNode = clusterService.localNode();
+        new ClusterStateObserver(clusterService, null, logger, threadContext).waitForNextChange(new RestoreListener(listener, localNode) {
```

Comment:
Does this not have the same issue as the original listener version, since the observer grabs the current version and we thus rely on one more cluster state update?

I think we can fix it by passing in `clusterService.state().version()-1` to the observer in the constructor.

Reply:
Alright @henningandersen as discussed elsewhere, I pushed https://github.com/elastic/elasticsearch/pull/96662/commits/e1bbe9a8a474a8b7e29660757aef873f1b838843 now which more safely checks the current state using the pattern we also use when working around this issue in tests.
See `org.elasticsearch.test.ClusterServiceUtils#awaitClusterState`.

Maybe instead of adding a tricky test here a follow-up that would extracting the check-before-wait logic so it can be used in a couple of spots and adding a test there would be an alternative? (promise I'd get to it this week ;))